debian/ilcheck.1

.\" Copyright (c) 2001 Southern Storm Software, Pty Ltd.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.TH PNET SECTION "October 21, 2001"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
ilverify \- IL Internal Call Check
.SH SYNOPSIS.ll +8
.B ilasm
[ options ]
.I "input.il"
\&...
.SH DESCRIPTION
This manual page documents briefly the
.B ilcheck
command.
This manual page was written for the Debian GNU/Linux distribution
because the original program does not have a manual page.
Instead, it has documentation in the GNU Info format; see below.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBpnet\fP is a program that...
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
.BR bar (1),
.BR baz (1).
.br
The programs are documented fully by
.IR "The Rise and Fall of a Fooish Bar" ,
available via the Info system.
.SH AUTHOR
This manual page was written by Andrew Mitchell <ajmitch@ajmitch.dhis.org>,
for the Debian GNU/Linux system (but may be used by others).


NEWS

0.6.12 (16 January 2005)

	Runtime engine:

	* Beginning of PPC unroller implementation (Gopal V).

	C# Compiler:

	* Compatibility with /debug flags in csc (Gopal V).
	* Handling of method/property/etc declarations that hide base
	  class virtuals and access checks for same (Richard Baumann).

	Documentation:

	* Information on debugging the unroller with gdb (Gopal V).

	Platform Support:

	* Use a portable version of tparm to match Susv3 specs (Gopal V).
	* Case sensitivity during comparisons (Klaus Treichel).
	* EAX vs REG_EAX usage in interrupt_posix (Gopal V).
	* Fix cache flush logic for PPC (Gopal V).
	* Skip empty lines in "@file" response files (Stephen Compall).

0.6.10 (30 October 2004)

	Runtime engine:

	* Null parameter names when emitting parameters (Gopal V).
	* Various fixes to System.Reflection internalcalls (Gopal V).
	* Thread suspend/resume (Thong Nguyen).
	* Internalcalls for CustomAttributeBuilder (Rhys Weatherley).
	* Use "fprem", not "fprem1" for x86 IL remainder (Rhys Weatherley).
	* Handle vararg methods and delegates more centrally (Gopal V).
	* System.Decimal division bug (Klaus Treichel).
	* Do not inline loads from value types or crashes will occur (Gopal V).

	C# Compiler:

	* Work around some PPC64-related optimization problesm (Gopal V).
	* Nested class "famorassem" handling (Gopal V).
	* Fixes to enum initializers (Gopal V).

	BF Compiler:

	* Fix the behavior of BFOptimize() which currently destroys the
	  parse tree (Gopal V).

	Loader and Metadata:

	* Prevent double loading of assemblies by using the full path (Gopal V).
	* Use ILNativeUInt instead of "unsigned long" in il_types.h, which
	  fixes some problems with 64-bit platforms (Rhys Weatherley).
	* Properly import types before writing them out with SR.Emit (Gopal V).
	* Add a flag to enable ignoring load errors (Rhys Weatherley).

	Assembler:

	* Forward references for data labels (Gopal V).

	Platform Support:

	* Get user name via uid, rather than getlogin (Carl-Adam Brengesjo).
	* Enable threading for Solaris (Robin Kay).
	* Fix boot time calculation at the nanosecond level (Gopal V).
	* Bug in socket selection for infinite selects (Gopal V).
	* Alter gen_errno.sh to handle errno codes that are split across
	  multiple header files (Rhys Weatherley).
	* Response file support in cpp for systems with limited command
	  line lengths (Maciek Plewa).
	* Socket flags for Solaris (Adam J. Tybor).
	* Port socket code to hpux by including <sys/modem.h> (Gopal V).
	* Fix daylight savings for Win32 (Carl-Adam Brengesjo).
	* Libffi bug in x86-64, causing incorrect stack alignment (Gopal V).
	* Serial I/O routines for Win32 (Gopal V).
	* Timeouts in Unix version of serial reads (Adam Gibson).
	* Various profile fixes (Gopal V).
	* BeOS port (Tomas Wilhelmsson).
	* Cris port (Simon Posnjak).
	* Man page formatting problems (Peter Lund).

	Other:

	* Clean up Debian packaging (Russell Stuart).

0.6.8 (8 August 2004)

	Runtime engine:

	* Module.GetTypes and Assembly.GetExportedTypes internalcall (Gopal V).
	* Fixes to IL offset determination in the method cache (Gopal V).
	* "synchronized" method handling (Thong Nguyen).
	* Be careful not to notify finalizers during allocation (Thong Nguyen).
	* Inlining of Math functions (Thong Nguyen).
	* Inlining simple properties that get/set fields (Thong Nguyen).
	* Various fixes in the finalizer code (Thong Nguyen).
	* Thread aborting (Thong Nguyen).
	* Using signals to detect NULL references and the like (Thong Nguyen).
	* Thread execution contexts (Thong Nguyen).
	* Dns and Socket internalcalls (Gopal V).
	* GCAddrOfPinnedObject fixes (Llewellyn Pritchard).
	* Extended console support (Rhys Weatherley).

	C# Compiler:

	* Correct line numbers for argumentvar nodes (Gopal V).
	* Fixes to identifier resolution (Gopal V).
	* Coercion fixes in the arguments to user-defined operators (Gopal V).
	* Create nested classes in the proper order (Gopal V).
	* Null values and IntPtr (Gopal V).
	* Fixes for indexers on value types (Gopal V).
	* Assign add operations involving strings and objects (Gopal V).
	* Standalone assembly attributes (Rhys Weatherley).
	* Resolve classes before allocating heap objects (Gopal V).
	* Fix do ... while grammar to require a terminating ';' (Gopal V).
	* Add the "-fhidebyname" and "-fhidebysig" options (Rhys Weatherley).
	* Error message for private interface modifiers (Gopal V).
	* Check for event types before performing add/remove operations (Gopal V).

	C Compiler:

	* Fix ILNode_GetType for C strings (Gopal V).
	* Fix semantic analysis for member accesses (Gopal V).

	Loader and Metadata:

	* CompareUnicode not working correctly (Thong Nguyen).
	* ILClassImplements on a class should return true for itself (T. Nguyen).
	* Serialization of null strings in attribute blocks (Rhys Weatherley).
	* Recognise "OptionalAttribute" and "DefaultValueAttribute" (Rhys W).
	* Resolve classes in "TypeAssignCompatible" (Gopal V).
	* Recognise generic classes in files generated by .NET 2.0 (Rhys W).

	Linker:

	* Resource linking problems (Llewellyn Pritchard).
	* Obey "AssemblyVersionAttribute" if it is present (Rhys Weatherley).
	* "-fmetadata-version" option (Rhys Weatherley).
	* "-fculture" option (Rhys Weatherley).

	Platform Support:

	* Compilation fixes for 64-bit (Andrzej Dopierala).
	* Utilities for lists and queues (Thong Nguyen).
	* Various windows build fixes (Thong Nguyen).
	* ILGetSinceRebootTime for non-Windows platforms (Thong Nguyen).
	* Faster versions of volatile and interlocked functions (Thong Nguyen).
	* Daylight savings problem (David Logan).
	* Deal with trailing whitespace in response files (Rhys Weatherley).
	* OpenBSD support (Saul Hazledine).
	* Support for parallel builds (Gopal V).
	* Fix interlock functions for amd64 (Gopal V).

	Documentation:

	* Update the HACKING file (Klaus Treichel).
	* Typo in the ilgac manual (Richard Baumann).

0.6.6 (29 May 2004)

	Runtime engine:


ilnative/ilcheck.c
/*
 * ilcheck.c Check runtime calls against the mscorlib.dll
 *
 * Copyright (C) 2001  Southern Storm Software, Pty Ltd.
 *
 * Hacked by Gopal.V from ilnative.c
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdio.h>
#include "il_system.h"
#include "il_image.h"
#include "il_dumpasm.h"
#include "il_utils.h"
#include "il_engine.h"
#include "../engine/engine.h"

typedef struct _ILDiffDict
{
	char *name;
	void *info;
	struct _ILDiffDict *methods;
	struct _ILDiffDict *prev;
}ILDict;

ILDict * ILDict_new(ILDict *prev,char *name)
{
	ILDict *retval=(ILDict*)calloc(1,sizeof(ILDict));
	if(name!=NULL)retval->name=name;
	retval->prev=prev;
	return retval;
}

ILDict * ILDict_search(ILDict *end,char *name)
{
	ILDict *i;
	for(i=end;i!=NULL;i=i->prev)
	{
		if(!strcmp(i->name,name))return i;
	}
	return NULL;
}

ILDict *klasses;
#ifdef	__cplusplus
extern	"C" {
#endif

/*
 * Table of command-line options.
 */
static ILCmdLineOption const options[] = {
	{"-v", 'v', 0, 0, 0},
	{"-a",'a',0,"-a","output all internal calls"},
	{"--all",'a',0,0,0},
	{"-fxml",'f',1,"-fxml","output XML instead of human-readable text"},
	{"--version", 'v', 0,
		"--version    or -v",
		"Print the version of the program."},
	{"--help", 'h', 0,
		"--help",
		"Print this help message."},
	{0, 0, 0, 0, 0}
};

static void usage(const char *progname);
static void version(void);
static int printNatives(const char *filename, ILContext *context,
						int xml,int all);

int main(int argc, char *argv[])
{
	char *progname = argv[0];
	int sawStdin;
	int state, opt;
	char *param;
	int errors,xml=0,all=0;
	ILContext *context;

	/* Parse the command-line arguments */
	state = 0;
	while((opt = ILCmdLineNextOption(&argc, &argv, &state,
									 options, &param)) != 0)
	{
		switch(opt)
		{
			case 'v':
			{
				version();
				return 0;
			}
			/* Not reached */
			case 'f':
			{
				xml=1;
				break;
			}
			case 'a':
			{
				all=1;
				break;
			}
			default:
			{
				usage(progname);
				return 1;
			}
			/* Not reached */
		}
	}

	/* We need at least one input file argument */
	if(argc <= 1)
	{
		usage(progname);
		return 1;
	}

	/* Create a context to use for image loading */
	context = ILContextCreate();
	if(!context)
	{
		fprintf(stderr, "%s: out of memory\n", progname);
		return 1;
	}

	/* Load and print information about the input files */
	sawStdin = 0;
	errors = 0;
	while(argc > 1)
	{
		if(!strcmp(argv[1], "-"))
		{
			/* Dump the contents of stdin, but only once */
			if(!sawStdin)
			{
				errors |= printNatives("-", context, xml,all);
				sawStdin = 1;
			}
		}
		else
		{
			/* Dump the contents of a regular file */
			errors |= printNatives(argv[1], context, xml,all);
		}
		++argv;
		--argc;
	}

	/* Destroy the context */
	ILContextDestroy(context);
	
	/* Done */
	return errors;
}

static void usage(const char *progname)
{
	fprintf(stdout, "ILCHECK " VERSION " - IL Internal Call Check \n");
	fprintf(stdout, "Copyright (c) 2001 Southern Storm Software, Pty Ltd.\n");
	fprintf(stdout, "          (c) 2002 FSF India.\n");
	fprintf(stdout, "\n");
	fprintf(stdout, "Usage: %s [options] input ...\n", progname);
	fprintf(stdout, "\n");
	ILCmdLineHelp(options);
}

static void version(void)
{

	printf("ILCHECK " VERSION " - IL Internal Call Check \n");
	printf("Copyright (c) 2001 Southern Storm Software, Pty Ltd.\n");
	printf("          (c) 2002 FSF India.\n");
	printf("\n");
	printf("ILCHECK comes with ABSOLUTELY NO WARRANTY.  This is free software,\n");
	printf("and you are welcome to redistribute it under the terms of the\n");
	printf("GNU General Public License.  See the file COPYING for further details.\n");
	printf("\n");
	printf("Use the `--help' option to get help on the command-line options.\n");
}

/*
 * Dump information about a native method.
 */
static void dumpMethodInfoTxt(ILImage *image, ILMethod *method,char *stat)
{
	/* Dump the method attributes */
	ILDumpFlags(stdout, ILMethod_Attrs(method),
				ILMethodDefinitionFlags, 0);

	/* Dump the method signature */
	ILDumpMethodType(stdout, image, ILMethod_Signature(method), 0,
					 ILMethod_Owner(method), ILMethod_Name(method),
					 method);
	putc(' ', stdout);

	/* Dump the implementation flags */
	ILDumpFlags(stdout, ILMethod_ImplAttrs(method),
				ILMethodImplementationFlags, 0);

	/* Terminate the line */
	putc('\n', stdout);
}

static void dumpMethodInfoXml(ILImage *image, ILMethod *method,char *stat)
{
	/* Dump the method attributes */
	printf("\t <internalcall name=\"%s\"",ILMethod_Name(method));
	printf("class = \"%s.%s\" ",ILClass_Namespace(ILMethod_Owner(method)),ILClass_Name(ILMethod_Owner(method)));
	printf(" signature=\"");
	/* Dump the method signature */
	ILDumpMethodType(stdout, image, ILMethod_Signature(method), 0,
					 ILMethod_Owner(method), ILMethod_Name(method),
					 method);
	/* Terminate the line */
	printf("\" status=\"%s\"/>\n",stat);
}
static void addmethodinfo(ILImage *image, ILMethod *method,char *stat)
{
	/* Dump the method attributes */
	char *name=(char*)ILClass_Name(ILMethod_Owner(method));
	ILDict *this_class;	
	if(!ILDict_search(klasses,name))
	{	
		klasses=ILDict_new(klasses,name);
	}
	this_class=ILDict_search(klasses,name);
	this_class->methods=ILDict_new(this_class->methods,stat);
	this_class->methods->info=method;
}
/*
 * Load an IL image an display the native methods.
 */
static int printNatives(const char *filename, ILContext *context,
						int xml,int all)
{
	ILImage *image;
	unsigned long numMethods;
	unsigned long token;
	ILMethod *method;
	
	ILDict *klass,*mthd;
	
	klasses=ILDict_new(NULL,"<start>");
	/* Attempt to load the image into memory */
	if(ILImageLoadFromFile(filename, context, &image,
					  	   IL_LOADFLAG_FORCE_32BIT |
					  	   IL_LOADFLAG_NO_RESOLVE, 1) != 0)
	{
		return 1;
	}

	/* Walk the MethodDef table and print all internalcalls that are
	   missing from the internalcall table */
	numMethods = ILImageNumTokens(image, IL_META_TOKEN_METHOD_DEF);
	for(token = 1; token <= numMethods; ++token)
	{
		method = (ILMethod *)ILImageTokenInfo
									(image, IL_META_TOKEN_METHOD_DEF | token);
		if(method)
		{
			if(ILMethod_IsInternalCall(method) ||
			        ILMethod_IsNative(method))
			{								
				if(!_ILFindInternalCall(method,0))
				{
					addmethodinfo(image,method,"MISSING");
				}
				else if(all)
				{
					addmethodinfo(image,method,"OK");
				}
			}
		}
	}
	if(xml)printf("<InternalCallStatus>\n");
	for(klass=klasses;klass->prev!=NULL;klass=klass->prev)
	{
		if(xml)
		{
			printf("<class namespace=\"%s\" name=\"%s\">\n",ILClass_Namespace(ILMethod_Owner(klass->methods->info)),klass->name);
		}
		for(mthd=klass->methods;mthd!=NULL;mthd=mthd->prev)
		{
			if(xml)dumpMethodInfoXml(image,mthd->info,mthd->name);
			else dumpMethodInfoTxt(image,mthd->info,mthd->name);
		}
		if(xml)
		{
			printf("</class>\n");
		}
	}
	if(xml)printf("</InternalCallStatus>\n");
	/* Clean up and exit */
	ILImageDestroy(image);
	return 0;
}

#ifdef	__cplusplus
};
#endif
